cmd/snap-bootstrap/triggerwatch/evdev.go

```diff
@@ -0,0 +1,227 @@
+	return isDown, nil
+}
+
+func (e *evdevKeyboardInputDevice) WaitForTrigger(ch chan keyEvent) {
+	logger.Noticef("%s: starting wait", e)
+
+	// XXX: do not mess with setting the key repeat rate, as it's cumbersome
+	// and golang-evdev SetRepeatRate() parameter order is actually reversed
+	// wrt. what the kernel does. The evdev interprets EVIOCSREP arguments
+	// as (delay, repeat)
+	// https://elixir.bootlin.com/linux/latest/source/drivers/input/evdev.c#L1072
+	// but the wrapper is passing is passing (repeat, delay)
+	// https://github.com/gvalkov/golang-evdev/blob/287e62b94bcb850ab42e711bd74b2875da83af2c/device.go#L226-L230
+
+	holdState, err := e.probeKeyState()
+	if err != nil {
+		ch <- keyEvent{Err: fmt.Errorf("cannot obtain initial key state: %v", err), Dev: e}
+	}
+	if holdState {
+		// looks like the key is pressed initially, we don't know when
+		// that happened, but prentend got pressed now
+		logger.Noticef("%s: key is already down", e)
+		// XXX: consider shortening the hold time?
+	}
+
+	type evdevEvent struct {
+		kev *evdev.KeyEvent
+		err error
+	}
+
+	// buffer large enough to collect some events
+	evChan := make(chan evdevEvent, 10)
+
+	monitorKey := func() {
+		for {
+			ies, err := e.dev.Read()
+			if err != nil {
+				evChan <- evdevEvent{err: err}
+				break
+			}
+			for _, ie := range ies {
+				if ie.Type != evdev.EV_KEY || ie.Code != e.keyCode {
+					continue
+				}
+				kev := evdev.NewKeyEvent(&ie)
+				evChan <- evdevEvent{kev: kev}
+			}
+		}
+		close(evChan)
+	}
+
+	go monitorKey()
+
+	holdTimer := time.NewTimer(holdToTrigger)
+	// no sense to keep it running later either
+	defer holdTimer.Stop()
+
+	if !holdState {
+		// key isn't held yet, stop the timer
+		holdTimer.Stop()
+	}
+Loop:
+	for {
+		select {
+		case ev := <-evChan:
+			if ev.err != nil {
+				holdTimer.Stop()
+				ch <- keyEvent{Err: err, Dev: e}
+				break Loop
+			}
+			kev := ev.kev
+			switch kev.State {
+			case evdev.KeyDown:
+				holdState = true
+				holdTimer.Reset(holdToTrigger)
+				logger.Noticef("%s: trigger key down", e)
+			case evdev.KeyHold:
+				if !holdState {
+					holdState = true
+					holdTimer.Reset(holdToTrigger)
+					logger.Noticef("%s: unexpected hold without down", e)
+				}
+			case evdev.KeyUp:
+				holdTimer.Stop()
+				holdState = false
+				logger.Noticef("%s: trigger key up", e)
+			}
+		case <-holdTimer.C:
+			logger.Noticef("%s: hold complete", e)
+			ch <- keyEvent{Dev: e}
+			break Loop
+		}
+	}
+}
+
+func (e *evdevKeyboardInputDevice) String() string {
+	return fmt.Sprintf("%s: %s", e.dev.Phys, e.dev.Name)
+}
+
+func (e *evdevKeyboardInputDevice) Close() {
+	e.dev.File.Close()
+}
+
+type evdevInput struct{}
+
+func (e *evdevInput) FindMatchingDevices(filter triggerEventFilter) ([]triggerDevice, error) {
+	devices, err := evdev.ListInputDevices()
+	if err != nil {
+		return nil, fmt.Errorf("cannot list input devices: %v", err)
+	}
+
+	// XXX: this is set up to support key input only
```

Comment:
what is this comment trying to convey? I'm not sure I understand

Reply:
Rephrased. We support key input devices only. There's actually more types of devices available under `/dev/input/*`, mice, touchpads, joysticks.